### PR TITLE
feat: unified command permissions framework

### PR DESCRIPTION
**Pull request by doosedev**
## Description

This PR removes all command-specific conditionals for checking role & channel permissions, and moves them along with existing permissions to a unified checker called by the command handler.

## Test Results

WIP

## Discord Username

pererry#8489

### CODE REVIEW

**Comment by benw202** [comment]
Liking this so far! Nice work

File: src/constants.ts [hunk 1]
```diff
@@ -77,3 +77,5 @@ export const AircraftTypeList = {
     a32nx: '32:1044695612337168494',
     a380x: '38:1044695718348210177',
 };
+
+export const PermissionsEmbedDelay = 10;
```
Comment (pdellaert) [inline src/constants.ts:81]:
```suggestion
export const PermissionsEmbedDelay = 10;

```
Should always have a newline/linefeed at the end of a file.

File: src/commands/utils/ping.ts [hunk 2]
```diff
@@ -6,7 +6,7 @@ export const ping: CommandDefinition = {
     name: 'ping',
     description: 'Send back a message',
     category: CommandCategory.UTILS,
-    requiredPermissions: ['ManageWebhooks'],
+    requirements: { permissions: ['ManageWebhooks'] },
```
Comment (pdellaert) [inline src/commands/utils/ping.ts:9]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to BOT_DEVELOPER, ADMIN & MODERATOR roles?

File: src/commands/moderation/welcome.ts [hunk 3]
```diff
@@ -79,7 +79,7 @@ const HELP_EMBED = makeEmbed({
 export const welcome: CommandDefinition = {
     name: 'welcome',
     description: 'Sends the welcome',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/welcome.ts:82]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/warn/warn.ts [hunk 4]
```diff
@@ -100,7 +94,10 @@ const noModLogs = makeEmbed({
 
 export const warn: CommandDefinition = {
     name: 'warn',
-    requiredPermissions: ['BanMembers'],
+    requirements: {
+        permissions: ['BanMembers'],
```
Comment (pdellaert) [inline src/commands/moderation/warn/warn.ts:98]:
```suggestion
```
Not a comment on the PR, but @benw202 : I'd suggest removing the permissions, it really doesn't have much of a point with the roles there anyway.

File: src/commands/moderation/warn/listWarnings.ts [hunk 5]
```diff
@@ -17,15 +17,12 @@ const noConnEmbed = makeEmbed({
...
-    title: 'Warn',
-    description: 'You do not have permission to use this command.',
-    color: Colors.Red,
-});
-
 export const listWarnings: CommandDefinition = {
     name: ['warnings', 'listwarn', 'listwarnings', 'warns'],
-    requiredPermissions: ['BanMembers'],
+    requirements: {
+        permissions: ['BanMembers'],
```
Comment (pdellaert) [inline src/commands/moderation/warn/listWarnings.ts:23]:
```suggestion
```
Not a comment on the PR, but @benw202 : I'd suggest removing the permissions, it really doesn't have much of a point with the roles there anyway.

File: src/commands/moderation/warn/deleteWarn.ts [hunk 6]
```diff
@@ -49,7 +43,10 @@ const deleteEmbed = makeEmbed({
 
 export const deleteWarn: CommandDefinition = {
     name: ['deletewarn', 'delwarn', 'deletewarning'],
-    requiredPermissions: ['BanMembers'],
+    requirements: {
+        permissions: ['BanMembers'],
```
Comment (pdellaert) [inline src/commands/moderation/warn/deleteWarn.ts:47]:
```suggestion
```
Not a comment on the PR, but @benw202 : I'd suggest removing the permissions, it really doesn't have much of a point with the roles there anyway.

File: src/commands/moderation/untimeout.ts [hunk 7]
```diff
@@ -67,7 +67,7 @@ const failedUnTimeoutEmbed = (user: User) => (makeEmbed({
 
 export const untimeout: CommandDefinition = {
     name: ['untimeout', 'removetimeout'],
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/untimeout.ts:70]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/unban.ts [hunk 8]
```diff
@@ -7,7 +7,7 @@ type UserLike = User | Snowflake
 
 export const unban: CommandDefinition = {
     name: 'unban',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/unban.ts:10]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/timeout.ts [hunk 9]
```diff
@@ -149,7 +149,7 @@ const failedTimeoutEmbed = (user: User) => (makeEmbed({
 
 export const timeout: CommandDefinition = {
     name: 'timeout',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/timeout.ts:152]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/rules.ts [hunk 10]
```diff
@@ -55,7 +55,7 @@ const ROLE_EMBED = makeEmbed({
 export const rules: CommandDefinition = {
     name: 'rules',
     description: 'Sends the rules',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/rules.ts:58]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/roleassignment.ts [hunk 11]
```diff
@@ -27,7 +27,7 @@ const MEDIA_ANNOUNCEMENT_EMBED = makeEmbed({
 export const roleassignment: CommandDefinition = {
     name: 'roleassignment',
     description: 'Sends the role assignment messages',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/roleassignment.ts:30]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/faq.ts [hunk 12]
```diff
@@ -82,7 +82,7 @@ const faqEmbeds = [
 export const faq: CommandDefinition = {
     name: 'faq',
     description: 'Sends the FAQ',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/faq.ts:85]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

File: src/commands/moderation/ban.ts [hunk 13]
```diff
@@ -129,7 +129,7 @@ const noDM = (user: User) => makeEmbed({
 
 export const ban: CommandDefinition = {
     name: 'ban',
-    requiredPermissions: ['BanMembers'],
+    requirements: { permissions: ['BanMembers'] },
```
Comment (pdellaert) [inline src/commands/moderation/ban.ts:132]: Not a comment on the PR, but @benw202 : Does this make sense, should we take the opportunity to change this to ADMIN & MODERATOR roles?

This is, perhaps with unban, the only one that makes a bit of sense :)

**Review by pdellaert — CHANGES_REQUESTED** [review]
Great solution for permissions! Think I found one bug (before testing), but otherwise I have no comment on the code at this time. (All the comments you see about using the opportunity to clean up some old permissions)

It would be good to also add a section to the `CONTRIBUTING.md` on all the different permissions that can be configured for commands. Documenting the different Permission objects. Something like (I wrote this to make sure I understand the purpose myself, for testing):

* `permissions`: Optional list of all the discord permissions a user need to have.
* `permissionsError`: The message returned to the user in case they do not have all of the `permissions`. If not provided, a generic message will be returned.
* `roles`: Optional list of `Roles` the user must have at least one of. The roles that can be provided, are defined in the `constants.ts` file. 
* `rolesError`: The message returned to the user in case they do not have any of the `roles`, or if they have at least one of the `roles` in case `rolesBlacklist` is `true`. If not provided, a generic message will be returned.
* `rolesBlacklist`: Boolean to indicate if the list needs to be handled as a blacklist instead of the default whitelist behavior.
* `channels`: a list of `Channels` or `Threads` to identify in which Discord Channels or Threads this command can be executed, or is banned from being executed in case `channelsBlacklist` is set to `true`.
* `channelsError`: The message returned to the user in case the command is executed in a channel not in the `channels` list, or if it is executed in a channel that is in the `channels` list in case `channelsBlacklist` is `true`. If not provided, a generic message will be returned.
* `channelsBlacklist`: Boolean to indicate if the list needs to be handled as a blacklist instead of the default whitelist behavior.
* `quietErrors`: Boolean to indicate if permission errors should not result in a message to the user and instead should fail silently.
* `verboseErrors`: Boolean to indicate that if no specific message is provided through `permissionsError`, `rolesError` and/or `channelsError` that a verbose explanation should be provided instead of the default simple message.

But that could be done after merging this PR as well. 

_I have not done extensive testing, will do so in a follow up._

File: src/lib/command.ts [hunk 14]
```diff
@@ -25,3 +39,99 @@ export function isExecutorCommand(command: BaseCommandDefinition) {
...
+
+    // all checks have passed, we're cleared for liftoff
+    return [true, ''];
+}
+
+export async function sendPermissionsEmbed(msg: Message, error: string) {
+    // return silently for quiet error commands
+    if (!msg) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (doosedev) [reply src/lib/command.ts:126]: great catch, i must've missed this by accidental combination of the 10 millisecond delete timer and this bug 😅

File: src/lib/command.ts [hunk 15]
```diff
@@ -25,3 +39,99 @@ export function isExecutorCommand(command: BaseCommandDefinition) {
...
+            return [false, ''];
+        } if (requirements.rolesError) {
+            return [false, requirements.rolesError];
+        } if (requirements.verboseErrors) {
+            let errorText: string;
+
+            if (requirements.rolesBlacklist) {
+                errorText = `The ${requirements.roles.map((r) => member.guild.roles.cache.get(r).name).join(', ')} role${requirements.roles.length > 1 ? 's are' : ' is'} not allowed to use that!`;
+            } else {
+                errorText = `Only the ${requirements.roles.map((r) => member.guild.roles.cache.get(r).name).join(', ')} role${requirements.roles.length > 1 ? 's are' : ' is'} are allowed to use that!`;
```
Comment (pdellaert) [inline src/lib/command.ts:90]:
```suggestion
                errorText = `Only the ${requirements.roles.map((r) => member.guild.roles.cache.get(r).name).join(', ')} role${requirements.roles.length > 1 ? 's are' : ' is'} allowed to use that!`;
```

**Review by pdellaert — CHANGES_REQUESTED** [review]
Testing showed a bug in language, which isn't in the test report itself as the test focussed on permissions itself.
Screenshot 2023-02-04 at 10 26 30 PM
Screenshot 2023-02-04 at 10 30 42 PM

✅ **275/275 tests passed** ✅ 
**Full test report:** 
https://docs.google.com/spreadsheets/d/1tB-z46PuOj3ARvhCOAuow2f9w1BG2jxZzDoIByoRUYk/edit?usp=sharing

**Comment by benw202** [comment]
@pdellaert - Agree on your comments about changing perms. Like Pererry said on Discord, I'm happy for these to be done in another PR.

@doosedev - I'm also testing today. I would agree with Straks that an entry to `contributing.md` should be added. I'd like it added in this PR please so that new/existing contribs can grab a quick understanding of new functionality and the different options within it. (Like blacklist, the different types of errors it can spit out etc.)

File: .github/CONTRIBUTING.md [hunk 16]
```diff
@@ -232,3 +232,58 @@ export const choices: MessageCommandDefinition = {
...
+| rolesError        | Custom error message to return for failed role checks. If omitted, a generic error will be returned. |
+| rolesBlacklist    | Boolean flag to indicate whether the list of `roles` should be treated as a blacklist (defaults as whitelist). |
+| channels          | List of `Channels` & `Threads` that the command can be executed in, or is not allowed to be executed in if blacklisted. |
+| channelsError     | Custom error message to return for failed channel checks. If omitted, a generic error will be returned. |
+| channelsBlacklist | Boolean flag to indicate whether the list of `channels` should be treated as a blacklist (defaults as whitelist). |
+| quietErrors       | Boolean flag to disable role messages completely & silently fail out of command execution. Will override all other error message fields. |
+| verboseErrors     | Boolean flag to enable a more verbose error message for failed requirement checks. Any `permissionsError`, `rolesError`, or `channelsError` will override this setting in their respective groups. |
```
Comment (pdellaert) [inline .github/CONTRIBUTING.md:289]:
```suggestion
| Field             | Description                                                                                                                                                                                        |
|:------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| permissions       | List of all Discord type permissions a member must have.                                                                                                                                           |
| permissionsError  | Custom error message to return for a failed permission checks. If omitted, a generic error will be returned.                                                                                       |
| roles             | List of `Roles` which a member must have at least one of, or must not have any of if blacklisted. Roles are defined in `constants.ts`.                                                             |
| rolesError        | Custom error message to return for a failed role checks. If omitted, a generic error will be returned.                                                                                             |
| rolesBlacklist    | Boolean flag to indicate whether the list of `roles` should be treated as a blacklist (defaults as whitelist).                                                                                     |
| channels          | List of `Channels` & `Threads` that the command can be executed in, or is not allowed to be executed in if blacklisted. Channels are defined in `constants.ts`.                                    |
| channelsError     | Custom error message to return for a failed channel checks. If omitted, a generic error will be returned.                                                                                          |
| channelsBlacklist | Boolean flag to indicate whether the list of `channels` should be treated as a blacklist (defaults as whitelist).                                                                                  |
| quietErrors       | Boolean flag to disable permission failure messages completely & silently fail out of command execution. Will override all other error message fields.                                             |
| verboseErrors     | Boolean flag to enable a more verbose error message for failed requirement checks. Any `permissionsError`, `rolesError`, or `channelsError` will override this setting in their respective groups. |

```
Minor edits in text (just personal suggestions). Formatting is just me being a little neurotic (feel free to ignore). And newline at the end of the file.

**Review by pdellaert — COMMENTED** [review]
Minor tweaks on the doc, but feel free to ignore those if you think it's too much.

**Comment by benw202** [comment]
Have tested and like what has been done. I agree with the above requested changes. Give me a ping when resolved and I can approve and merge :)

**Review by pdellaert — APPROVED** [review]

**Review by pdellaert — APPROVED** [review]

**Review by benw202 — APPROVED** [review]

**Pull request merged**